src/components/guideLineLayer.ts

```diff
@@ -0,0 +1,174 @@
+///<reference path="../reference.ts" />
+
+module Plottable {
+export module Components {
+  export class GuideLineLayer<D> extends Component {
+    public static ORIENTATION_VERTICAL = "vertical";
+    public static ORIENTATION_HORIZONTAL = "horizontal";
+
+    private _orientation: string;
+    private _value: D;
+    private _scale: QuantitativeScale<D>;
+    private _pixelPosition: number;
+    private _scaleUpdateCallback: ScaleCallback<QuantitativeScale<D>>;
+    private _guideLine: d3.Selection<void>;
+
+    constructor(orientation: string) {
+      super();
+      if (orientation !== GuideLineLayer.ORIENTATION_VERTICAL && orientation !== GuideLineLayer.ORIENTATION_HORIZONTAL) {
+        throw new Error(orientation + " is not a valid orientation for GuideLineLayer");
+      }
+      this._orientation = orientation;
+      this._clipPathEnabled = true;
+      this.addClass("guide-line-layer");
+      this._scaleUpdateCallback = () => {
+        this._syncValueAndPixelPosition();
+        this.render();
+      };
+    }
+
+    protected _setup() {
+      super._setup();
+      this._guideLine = this.content().append("line").classed("guide-line", true);
+    }
+
+    protected _sizeFromOffer(availableWidth: number, availableHeight: number) {
+      return {
+        width: availableWidth,
+        height: availableHeight
+      };
+    }
+
+    private _isVertical() {
+      return this._orientation === GuideLineLayer.ORIENTATION_VERTICAL;
+    }
+
+    public fixedWidth() {
+      return true;
+    }
+
+    public fixedHeight() {
+      return true;
+    }
+
+    public computeLayout(origin?: Point, availableWidth?: number, availableHeight?: number) {
+      super.computeLayout(origin, availableWidth, availableHeight);
+      if (this.scale() != null) {
+        if (this._isVertical()) {
+          this.scale().range([0, this.width()]);
+        } else {
+          this.scale().range([this.height(), 0]);
+        }
+      }
+      return this;
+    }
+
+    public renderImmediately() {
+      super.renderImmediately();
+      this._syncValueAndPixelPosition();
+      this._guideLine.attr({
+        x1: this._isVertical() ? this.pixelPosition() : 0,
+        y1: this._isVertical() ? 0 : this.pixelPosition(),
+        x2: this._isVertical() ? this.pixelPosition() : this.width(),
+        y2: this._isVertical() ? this.height() : this.pixelPosition()
+      });
+      return this;
+    }
+
+    private _syncValueAndPixelPosition() {
+      if (this.scale() != null) {
+        if (this.value() != null) {
+          this._pixelPosition = this.scale().scale(this.value());
+        } else if (this.pixelPosition() != null) {
+          this._value = this.scale().invert(this.pixelPosition());
+        }
+      }
```

Comment:
You can avoid some of the nesting here by having
```typescript
if (this.scale() == null) {
  return;
}
```

It just makes the following code easier to read